lib/AST/UnqualifiedLookup.cpp

```diff
@@ -547,12 +553,61 @@ void UnqualifiedLookupFactory::lookInASTScopes() {
                               Name, Loc, consumer);
 }
 
+void ASTScopeDeclConsumerForUnqualifiedLookup::maybeUpdateSelfDC(
+    VarDecl *var) {
+  // We have a binding named 'self'.
+  //
+  // There are three possibilities:
+  //
+  // 1) This binding is the 'self' parameter of a method,
+  // 2) This binding is a bona-fide 'self' capture, meaning a capture
+  //    list entry named 'self' with initial value expression 'self',
+  // 3) None of the above.
+  //
+  // How we handle these cases depends on whether we've already seen
+  // another 'self' binding.
+  if (candidateSelfDC == nullptr) {
+    // We haven't seen one yet, so record it.
+    if (var->isSelfParameter())
+      candidateSelfDC = var->getDeclContext();
+    else if (var->isSelfParamCapture())
+      candidateSelfDC = var->getParentCaptureList()->getClosureBody();
+  } else {
+    // If we see a binding named 'self' that is not a bona-fide
+    // 'self', we have to forget about the previous 'self' capture
+    // because it's not going to be the right one for accessing
+    // instance members of the innermost nominal type. Eg,
+    //
+    // class C {
+    //   func bar() {}
+    //   func foo() {
+    //     _ { [self=12] { [self] bar() } }
+    //   }
+    // }
+    //
+    // Instead, we're going to move on and look for the next-innermost
+    // 'self' binding.
+    if (!var->isSelfParameter() &&
+        !var->isSelfParamCapture())
+      candidateSelfDC = nullptr;
+  }
+}
+
 bool ASTScopeDeclConsumerForUnqualifiedLookup::consume(
     ArrayRef<ValueDecl *> values, DeclVisibilityKind vis,
     NullablePtr<DeclContext> baseDC) {
   for (auto *value: values) {
     if (factory.isOriginallyTypeLookup && !isa<TypeDecl>(value))
       continue;
+
+    // Try to resolve the base for unqualified instance member
+    // references. This is used by lookInMembers().
+    if (auto *var = dyn_cast<VarDecl>(value)) {
+      if (var->getName() == factory.Ctx.Id_self) {
+        maybeUpdateSelfDC(var);
```

Comment:
Now that you've refactored--thanks!--I can see more clearly what is going on with candidateSelfDC. Is it a fragile piece of mutable state? Could we pass a piece of immutable state through the lookup recursion instead? It seems to be a bit odd that a `consume` method has a side effect. What do you think?

Reply:
I was actually going to try to refactor the lookup to use a loop that walks the parents instead of recursing. This will allow us to implement the one-level lookup we need to do re-declaration checking for locals defined in the same scope. Maybe then the selfDC computation can go there and be totally explicit, instead of being part of consume().